File: sdk/rust/src/artifact/nodejs.rs
use crate::{
    api::artifact::ArtifactSystem::{
        Aarch64Darwin, Aarch64Linux, UnknownSystem, X8664Darwin, X8664Linux,
    },
    artifact::{step, system, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::{bail, Result};

/// Build-target `Node.js` JavaScript runtime, fetched as a prebuilt binary release.
#[derive(Default)]
pub struct NodeJS {}

impl NodeJS {
    /// Creates a builder for the pinned `Node.js` release.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Builds the `Node.js` artifact for the context's target system.
    ///
    /// # Errors
    ///
    /// Returns an error if the target system has no known `Node.js` release, or if
    /// registering the source or artifact with the build context fails.
    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "nodejs";

        let system = context.get_system();

        let source_target = match system {
            Aarch64Darwin => "darwin-arm64",
            Aarch64Linux => "linux-arm64",
            X8664Darwin => "darwin-x64",
            X8664Linux => "linux-x64",
            UnknownSystem => bail!("unsupported {name} system: {}", system.as_str_name()),
        };

        let source_version = "22.22.0";
        let source_path = format!(
            "https://sdk.vorpal.build/source/node-v{source_version}-{source_target}.tar.gz"
        );

        let source = ArtifactSource::new(name, source_path.as_str()).build();

        let step_script = format!(
            "cp -pr \"./source/{name}/node-v{source_version}-{source_target}/.\" \"$VORPAL_OUTPUT\""
        );
        let steps = vec![step::shell(context, &[], &[], step_script, &[]).await?];

        Artifact::new(name, steps, system::SYSTEMS)
            .with_aliases(vec![format!("{name}:{source_version}")])
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/nodejs.ts
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";
import { getSystemStr, SYSTEMS } from "../system.js";

/**
 * Builder for the Node.js runtime artifact.
 *
 * Mirrors Rust `NodeJS` struct in `sdk/rust/src/artifact/nodejs.rs`.
 * Downloads and extracts the official Node.js binary distribution.
 */
export class NodeJS {
  async build(context: ConfigContext): Promise<string> {
    const name = "nodejs";
    const system = getSystemStr(context.getSystem());

    let sourceTarget: string;

    switch (system) {
      case "aarch64-darwin":
        sourceTarget = "darwin-arm64";
        break;
      case "aarch64-linux":
        sourceTarget = "linux-arm64";
        break;
      case "x86_64-darwin":
        sourceTarget = "darwin-x64";
        break;
      case "x86_64-linux":
        sourceTarget = "linux-x64";
        break;
      default:
        throw new Error(`unsupported ${name} system: ${system}`);
    }

    const sourceVersion = "22.22.0";
    const sourcePath = `https://sdk.vorpal.build/source/node-v${sourceVersion}-${sourceTarget}.tar.gz`;

    const source = new ArtifactSource(name, sourcePath).build();

    const stepScript = `cp -pr "./source/${name}/node-v${sourceVersion}-${sourceTarget}/." "$VORPAL_OUTPUT"`;
    const steps = [await shell(context, [], [], stepScript, [])];

    return new Artifact(name, steps, SYSTEMS)
      .withAliases([`${name}:${sourceVersion}`])
      .withSources([source])
      .build(context);
  }
}
